packages/flutter/lib/src/widgets/two_dimensional_viewport.dart

```diff
@@ -1304,23 +1356,59 @@ abstract class RenderTwoDimensionalViewport extends RenderBox implements RenderA
 
   void _insertChild(RenderBox child, ChildVicinity slot) {
     assert(_debugTrackOrphans(newOrphan: _children[slot]));
+    assert(!_keepAliveBucket.containsValue(child));
     _children[slot] = child;
     adoptChild(child);
   }
 
   void _moveChild(RenderBox child, {required ChildVicinity from, required ChildVicinity to}) {
-    if (_children[from] == child) {
-      _children.remove(from);
+    final TwoDimensionalViewportParentData childParentData = parentDataOf(child);
+    if (!childParentData.keptAlive) {
+      if (_children[from] == child) {
+        _children.remove(from);
+      }
+      assert(_debugTrackOrphans(newOrphan: _children[to], noLongerOrphan: child));
+      _children[to] = child;
+      return;
+    }
+    // If the child in the bucket is not current child, that means someone has
+    // already moved and replaced current child, and we cannot remove this
+    // child.
```

Comment:
I think this may only possible if the earlier movechild move another child into this place. In this case this child will be in _debugDanglingKeepAlives. Maybe add an assert check here?

Reply:
For this and the comment below, this matches almost exactly to the implementation in RenderSliverMultiBoxAdaptor, should the same changes be made there? I traced this back to https://github.com/flutter/flutter/pull/31978 for context.